_posts/2023-11-20-unlocking-the-secrets-to-ingestion.md

```diff
@@ -0,0 +1,191 @@
+
+For ingest heavy workloads we recommend setting the JVM size to half of your available memory. There are two ways this can be set:
+
+1. JVM config file (tarball install): `/config/jvm.options`
+
+```bash
+# /config/jvm.options
+
+# Min JVM memory size
+- Xms16g
+
+# Max JVM memory size
+- Xmx16g
+```
+
+1. Environment variable (container based installs): `OPENSEARCH_JAVA_OPTS`
+
+```bash
+OPENSEARCH_JAVA_OPTS="-Xms16g -Xmx16g“
+```
+
+### Translog Flush
+
+A translog flush is when OpenSearch commits the items that have been ingested but not persisted to on-disk Lucene segments. By default this happens every time the translog reaches 512MB. Increasing the translog flush size reduces the frequency of these operations which are resource intensive. Additionally they create larger Lucene segments which merge less frequently saving even more resources for ingestion. To see your current translog statistics you can use the [node stats](https://opensearch.org/docs/latest/api-reference/nodes-apis/nodes-stats/) endpoint like so: `curl -XPOST "<cluster url>/<index name>/_stats/flush?pretty"`
+
+We recommend setting the translog flush size to around 25% of your available Java heap defined earlier. On a node with 16GB of memory we would set our heap to 50% of the available memory which would be 8GB for the heap. Then we would take 25% of the heap to set aside for the translog flushes or 2GB.
+
+```json
+POST index-name/_settings
+{
+	"index" {
+		"translog.flush_threshold_size": "2048MB"
+	}
+}
+```
+
+Small note, the translog is not in memory. We set the translog threshold  to 25% of our memory because when the translog flushes it will transfer the data from disk into memory to be compiled into a Lucene segment to be persisted to disk.
+
+# Sharding and Replication
+
+We talked a little bit about threading earlier and now we will dive a little bit more into mechanisms for making things happen in parallel. OpenSearch uses shards to partition indexes. Which shard an index goes to is based off a hash of it’s `_id`. There are primary shards and replica shards. Primary shards are responsible for ingestion and sending the data to the replicas.
+
+### Primary Shard Distribution
+
+We want to ensure we prioritize balancing our shards across nodes so the workload is evenly distributed. Without setting this we could end up with a situation like the one below where all our primary shards for end up on the same node. That would cause our Node 1 here to be completely saturated while the other nodes may be underutilized.
```

Comment:
```suggestion
We want to ensure we prioritize balancing our shards across nodes so that the workload is evenly distributed. Without setting this we could end up with a situation like the one below where all our primary shards end up on the same node. That would cause our Node 1 to be completely saturated while underutilizing the other nodes.
```